runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java

```diff
@@ -0,0 +1,381 @@
+	}
+
+	@Override
+	protected void startUp()
+	{
+		Widget widgetBankTitleBar = client.getWidget(WidgetInfo.BANK_TITLE_BAR);
+
+		if (widgetBankTitleBar == null || widgetBankTitleBar.isHidden())
+		{
+			return;
+		}
+
+		calculateBankWithThread();
+	}
+
+	@Subscribe
+	public void onMenuChanged(MenuOptionClicked event)
+	{
+		if (!bankActive)
+		{
+			return;
+		}
+
+		String strMenuOption = event.getMenuOption();
+
+		if (strMenuOption.contains("Withdraw") ||
+				strMenuOption.contains("Deposit") ||
+				strMenuOption.contains("View tab") ||
+				strMenuOption.contains("View all"))
+		{
+			stopThreads();
+			calculateBankWithThread();
+		}
+	}
+
+	@Subscribe
+	public void onWidgetHidden(WidgetHiddenChanged event)
+	{
+		// Below is several checks to hopefully get a bank opening/closing event.
+		// If there are better events to check for this, then this should be changed.
+		// Then we we can remove the sleep() in calculateBankThread()
+		Widget widget = event.getWidget();
+
+		int group = WidgetInfo.TO_GROUP(widget.getId());
+
+		// This seems to be the only event sent when closing the bank
+		if (bankActive && group == WidgetID.RESIZABLE_VIEWPORT_BOTTOM_LINE_GROUP_ID && event.isHidden())
+		{
+			log.debug("Bank closing?");
+			stopThreads();
+			return;
+		}
+
+		// This event gets called a lot, so check to also see if it's are already bankActive
+		// This still does not work great, because it still gets called twice, but the first time is just a dud.
+		if (!bankActive && group == WidgetID.BANK_GROUP_ID)
+		{
+			calculateBankWithThread();
+		}
+	}
+
+	private void stopThreads()
+	{
+		bankActive = false;
+		log.debug("Bank value stopping");
+
+		if (executor != null)
+		{
+			executor.shutdownNow();
+		}
+
+		if (bankThread != null)
+		{
+			bankThread.interrupt();
+		}
+	}
+
+	private void calculateBankWithThread()
+	{
+		bankActive = true;
+		log.debug("Bank value running");
+		bankThread = new Thread(this::calculateBank);
+		bankThread.start();
+	}
+
+	class ItemPriceAndAlch extends ItemPrice
+	{
+		@Getter
+		@Setter
+		private int highAlchPrice;
+
+		ItemPriceAndAlch(int gePrice, int highAlchPrice)
+		{
+			setPrice(gePrice);
+			this.highAlchPrice = highAlchPrice;
+		}
+	}
+
+	private final Object rsItemLock = new Object();
+
+	/**
+	 * Grabs the GE price, calculates the high alch price, then multiplies the quantity
+	 * @param item
+	 * @return
+	 */
+	private ItemPriceAndAlch getItemPrices(WidgetItem item)
+	{
+		if (COINS == item.getId())
+		{
+			return new ItemPriceAndAlch(item.getQuantity(), item.getQuantity());
+		}
+
+		ItemComposition itemComposition;
+
+		// Having this lock seems to not crash runescape. I'm not sure if it's because of this method call
+		// or if it's something else that this lock is slowing down. I believe it's because this is the only
+		// place that has a possibility of accessing runescape's code.
+		synchronized (rsItemLock)
+		{
+			itemComposition = itemManager.getItemComposition(item.getId());
```

Comment:
It is not safe to call this from any thread other than the game thread. I would gather all of these upfront.